python/meterpreter/meterpreter.py

```diff
@@ -1260,11 +1260,37 @@ def add_channel(self, channel):
         return idx
 
     def add_process(self, process):
-        idx = self.next_process_id
-        self.processes[idx] = process
-        debug_print('[*] added process id: ' + str(idx))
-        self.next_process_id += 1
-        return idx
+        if has_windll:
+            PROCESS_ALL_ACCESS = 0x1fffff
+            OpenProcess = ctypes.windll.kernel32.OpenProcess
+            OpenProcess.argtypes = [ctypes.c_ulong, ctypes.c_long, ctypes.c_ulong]
+            OpenProcess.restype = ctypes.c_void_p
+            handle = OpenProcess(PROCESS_ALL_ACCESS, False, process.pid)
+        else:
+            handle = self.next_process_id
+            self.next_process_id += 1
+        self.processes[handle] = process
+        debug_print('[*] added process id: ' + str(process.pid) + ', handle: ' + str(handle))
+        return handle
+
+    def close_process(self, proc_h_id):
+        proc_h = self.processes.pop(proc_h_id, None)
+        if not proc_h:
+            return False
+        for channel_id, channel in self.channels.items():
+            if not isinstance(channel, MeterpreterProcess):
+                continue
+            if not channel.proc_h is proc_h:
+                continue
+            if has_windll:
+                CloseHandle = ctypes.windll.kernel32.CloseHandle
+                CloseHandle.argtypes = [ctypes.c_void_p]
+                CloseHandle.restype = ctypes.c_long
+                CloseHandle(proc_h_id)
+            self.close_channel(channel_id)
+            break
+        debug_print('[*] closed and removed process id: ' + str(proc_h.pid) + ', handle: ' + str(proc_h_id))
+        return True
```

Comment:
We don't want the `CloseHandle` call to be in the loop where we are iterating over the channels because if it is it'll only be closed when the process was channelized. Moving it to after this block ensure it's always called and the channel is removed before the handle is closed.

```suggestion
    def close_process(self, proc_h_id):
        proc_h = self.processes.pop(proc_h_id, None)
        if not proc_h:
            return False
        for channel_id, channel in self.channels.items():
            if not isinstance(channel, MeterpreterProcess):
                continue
            if not channel.proc_h is proc_h:
                continue
            self.close_channel(channel_id)
            break
        if has_windll:
            CloseHandle = ctypes.windll.kernel32.CloseHandle
            CloseHandle.argtypes = [ctypes.c_void_p]
            CloseHandle.restype = ctypes.c_long
            CloseHandle(proc_h_id)
        debug_print('[*] closed and removed process id: ' + str(proc_h.pid) + ', handle: ' + str(proc_h_id))
        return True
```